routing_plugin.go

```diff
@@ -104,3 +120,433 @@ func ReleaseRoutingPlugin(ctx context.Context) {
+		// just return the error.
+		if _, ok := privateEdges[nodeID]; !ok {
+			return nil, err
+		}
+
+		err = nil
+		nodeInfo = &lndclient.NodeInfo{
+			Node: &lndclient.Node{
+				PubKey: nodeID,
+			},
+		}
+	}
+
+	for _, edge := range privateEdges[nodeID] {
+		nodeInfo.Channels = append(nodeInfo.Channels, *edge)
+		nodeInfo.TotalCapacity += edge.Capacity
+		nodeInfo.ChannelCount++
+	}
+
+	return nodeInfo, err
+}
+
+// Init will initialize the "low to high" routing plugin. It'll save the MC
+// state and also preinit the internal state of the routing plugin. When the
+// instance is released, the saved MC state can be restored.
+func (r *lowToHighRoutingPlugin) Init(ctx context.Context, target route.Vertex,
+	routeHints [][]zpay32.HopHint, amt btcutil.Amount) error {
+
+	// Prepare the private edges from the passed route hints.
+	privateEdges := buildPrivateChannels(routeHints, target)
+
+	// Save the original target as the current "exit" node: where
+	// our payments should flow forward.
+	exit := target
+
+	// Walk until the first fork (if there's any). This first
+	// fork will be where we're going to try to manipulate success
+	// probabilities to increasingly prefer more expensive edges.
+
+	// We track all visited peers, so we won't end up walking
+	// back and forth on graphs that have no forks.
+	visited := make(map[route.Vertex]struct{})
+
+	for {
+		targetNodeInfo, err := getNodeInfo(
+			ctx, r.lnd.Client, target, privateEdges,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Make sure we mark the current node as visited.
+		visited[target] = struct{}{}
+
+		// If the target node has only one or more channels but all
+		// connects to the same peer, then we need to walk further.
+		var (
+			peer  route.Vertex
+			peers []route.Vertex
+		)
+		for _, edge := range targetNodeInfo.Channels {
+			if target != edge.Node1 {
+				peer = edge.Node1
+			} else {
+				peer = edge.Node2
+			}
+
+			if _, ok := visited[peer]; !ok {
+				visited[peer] = struct{}{}
+				peers = append(peers, peer)
+			}
+		}
+
+		if len(peers) == 1 {
+			exit = target
+			target = peers[0]
+			continue
+		}
+
+		// If there are no more peers to visit then we can't use
+		// this routing plugin.
+		if len(peers) == 0 {
+			return ErrRoutingPluginNotApplicable
+		}
+
+		// Found the first fork to our target.
+		break
+	}
+
+	nodeInfo, err := getNodeInfo(ctx, r.lnd.Client, target, privateEdges)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Routing plugin target: '%v' %v", nodeInfo.Alias,
+		nodeInfo.PubKey.String())
+
+	// Gather node info (including channels) for the nodes we're
+	// interested in.
+	targetChanged := exit != target
+	nodes := make(map[route.Vertex]*lndclient.NodeInfo)
+	for _, edge := range nodeInfo.Channels {
+		// Skip edges to the exit node since from there the route to
+		// the invoice target is always constructed from the same hops.
+		if targetChanged && (edge.Node1 == exit ||
```

Comment:
Why do we need `targetChanged` in here? Wouldn't we always skip edges to the exit node?